### PR TITLE
feat: `Event::Reopen` support for macOS

### PR DESCRIPTION
**Pull request by I-Info**
- [x] Tested on all platforms changed
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

## Description

This event is typically used to handle dock icon clicking event on macOS and  uses [`applicationShouldHandleReopen:hasVisibleWindows:`] callback from AppKit Framework. 

## Use case

It will make it possible to keep application running in backgound when all windows are closed, and create new window after dock icon clicked (application reopen callback).

[`applicationShouldHandleReopen:hasVisibleWindows:`]: https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428638-applicationshouldhandlereopen

### CODE REVIEW

File: src/event.rs
Comment (madsmtm) [inline src/event.rs:267]: Maybe document it more in Winit terms?

```suggestion
    /// Usually, the user would expect you to create a new window if there isn't any.
```

File: src/event.rs
Comment (madsmtm) [inline src/event.rs:274]: The user already knows how many windows they have, I don't really see the purpose of this flag?

```suggestion
    Reopen,
```

File: src/event.rs
Comment (madsmtm) [inline src/event.rs:264]: Copied more precise text from the docs.

```suggestion
    /// callback, which is usually called whenever the Finder reactivates an already running
    /// application because the user double-clicked it again or used the dock to activate it.
```

File: examples/reopen.rs
Comment (madsmtm) [inline examples/reopen.rs:1]: I think I'll wait 'till after https://github.com/rust-windowing/winit/pull/3447, as that reworks how our examples are handled.

**Review by madsmtm — CHANGES_REQUESTED** [review]
I can see this being useful, thanks!

File: examples/reopen.rs
Comment (I-Info) [inline examples/reopen.rs:2]: - [ ] Merge this example after #3447

**Review by kchibisov — CHANGES_REQUESTED** [review]
nack for 0.30, should be good for 0.31 with traits.

**Review by madsmtm — APPROVED** [review]
> nack for 0.30, should be good for 0.31 with traits.

I'm not going to block this on 0.31, it's a simple change that we can do now, and generally enables better behaviour on macOS.

**Comment by kchibisov** [comment]
And what should we do with all of that on other platforms? The event is macOS specific, but it bleeds into other backend and people will demand its handling, while in reality such event doesn't exist for them.

You should never trade quality for short term wins, this is just straight reduces overall quality because the event is not tied to window in particular.

**Comment by madsmtm** [comment]
> And what should we do with all of that on other platforms? The event is macOS specific, but it bleeds into other backend and people will demand its handling, while in reality such event doesn't exist for them.

I think the event is well documented as "Unsupported", which we already have precedent for, I don't see the issue with adding another one.

In the _end_ it won't really matter, since we're going to be changing things for 0.31 anyways. But like I said, in the months until that happens, this is a useful feature.

**Review by daxpedda — APPROVED** [review]
I nominated this specifically because I wanted to discuss this "blocking it because its platform-specific" issue in general, but we didn't get to it unfortunately.

Personally I strongly disagree that this somehow reduces the quality of the Winit API, we already have a big amount of events that are not cross-platform, this doesn't change anything.

That said, I'm glad we are trying to fix/improve this issue in v0.31.

**Comment by kchibisov** [comment]
All `Event` stuff is cross platform and has special semantics. `WindowEvent` stuff has specifics, but not the `Event`.

As I said, the solution is already here even for 0.30, if you want I can write it myself for macOS.

**Comment by madsmtm** [comment]
Instead of exposing this through Winit, https://github.com/rust-windowing/winit/pull/3758 is going to allow listening to `NSApplicationDelegate` events outside of Winit using `objc2::declare_class!`, Objective-C, Swift, or similar.

**Pull request closed without merging**
